mautic/mautic
# Allow a contact to go through a campaign multiple times

**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | Y
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/1606
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR is dependent on and includes #5942.

This introduces the ability for contacts to go through campaigns multiple times. 

If a contact is removed and added back, the `rotation` column is incremented by 1. This column is kept in sync with the event logs for the contact. This way we never have to delete contact history (always available in the timeline and reports) but allows the contact to "start over" in the campaign. 

[//]: # ( As applicable: )

#### Steps to test this PR:
1. Create a campaign with a basic flow of send email -> opened email? -> do some stuff. At the end of the campaign,  add a "change campaigns" action to restart the campaign. Run some contacts through the campaign. Notice when they reach the end and restart the campaign, they'll go through it again. 
2. Manually remove the contacts and they should no longer be picked up as part of the campaign
3. Add them back and they'll start the campaign over
4. Build a campaign event report and include the "Rotation" column. All the repetitive history should be seen in the report. 

#### List deprecations along with the new alternative:
1. CampaignModel::addLead, addLead, removeLead and removeLeads have been replaced by the MembershipManager
